Add ClusterIP monitoring for revision backends

greghaynes opened the pull request:
This adds logic for the RevisionBackendsManager to prefer sending
requests to a clusterIP if one is available, otherwise fallback to
podIPs.

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Related-To: #3885

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
NONE
```

greghaynes commented:
This is WIP as theres one failing test and Id like to add a few more tests for the new behavior, but otherwise this should be good to go.

tcnghia commented on pkg/activator/net/helpers.go:
```diff
@@ -0,0 +1,53 @@
+	wantName := networking.ServicePortName(protocol)
+	for _, p := range svc.Spec.Ports {
+		if p.Name == wantName {
+			port = int(p.Port)
```
could `return int(p.Port)` from here and `return -1` outside.

tcnghia replied:
also could return (int, error) here instead of using `-1`

tcnghia commented on pkg/activator/net/helpers.go:
```diff
@@ -0,0 +1,53 @@
+		}
+
+		pvs := strconv.Itoa(int(portVal))
+		for _, addr := range es.Addresses {
```
You could nest this loop to line 20 to avoid the `continue` on line 27, also could do `pvs := strconv.Itoa(port.Port)` instead of assigning to `portVal`.

tcnghia commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,94 @@ type destHealth struct {
+	}
+
+	if len(svcList) == 0 {
+		return nil, fmt.Errorf("Found no private service for revisoin %v", rw.rev)
```
revision

tcnghia commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,94 @@ type destHealth struct {
+		return nil, fmt.Errorf("Found no private service for revisoin %v", rw.rev)
+	}
+
+	if len(svcList) > 1 {
```
could group both to 

'Expected exactly 1 private service, found %d'

when `len(svcList) != 1`

greghaynes replied:
fixed with switch

tcnghia commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,94 @@ type destHealth struct {
+}
+
+func (rw *revisionWatcher) getK8sPrivateService() (*corev1.Service, error) {
+	selector, err := labels.Parse(fmt.Sprintf("%s=%s,%s=%s",
```
can we use
```
metav1.LabelSelector{
	MatchLabels: map[string]string{
		serving.RevisionLabelKey: ...,
	}
}
```
to avoid parsing and `err` handling.

tcnghia commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,94 @@ type destHealth struct {
+	if err != nil {
+		return nil, err
+	}
+	svcList, err := rw.serviceLister.Services(rw.rev.Namespace).List(selector)
```
Could probably inline `selector` here to clarify the List criteria in one line.

tcnghia commented on pkg/activator/net/revision_backends.go:
```diff
@@ -161,7 +231,9 @@ func (rw *revisionWatcher) checkDests() {
 
+	rw.logger.Infof("Done probing, got healthStates %v", healthStates)
 	if !reflect.DeepEqual(rw.healthStates, healthStates) {
+		rw.logger.Infof("sending update")
```
There is no format string, so `Info` instead of `Infof`?   Probably need a better log message here too.

greghaynes replied:
This was leftover debugging, removed it altogether

mattmoor commented on pkg/activator/net/helpers.go:
```diff
@@ -0,0 +1,53 @@
+package activator
```
missing header

mattmoor commented on pkg/activator/net/helpers_test.go:
```diff
@@ -0,0 +1,112 @@
+package activator
```
ditto

taragu commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
+	} else if ok {
+		rw.healthyClusterDest = dest
+		rw.updateCh <- &RevisionDestsUpdate{Rev: rw.rev, ClusterIPDest: dest}
+		return
```
Should we update `rw.healthStates` before returning?

greghaynes replied:
yes! good find

taragu commented on pkg/activator/net/revision_backends.go:
```diff
@@ -201,7 +276,9 @@ type revisionWatcherCh struct {
 // RevisionBackendsManager listens to revision endpoints and keeps track of healthy
 // l4 dests which can be used to reach a revisoin
```
typo: `revisoin`

taragu commented on pkg/activator/net/revision_backends.go:
```diff
@@ -267,14 +347,14 @@ func (rbm *RevisionBackendsManager) getOrCreateDestsCh(rev RevisionID) chan []st
 }
 
 // deleteRevisionWatcher deletes the revision wathcher for rev if it exists. It expects
 // a write lock is held on revisoinWatchersMux when calling.
```
typo: `revisoinWatchersMux`

greghaynes commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
+		rw.logger.Errorw("Failed to probe clusterIP", zap.Error(err))
+	} else if ok {
+		rw.healthyClusterDest = dest
+		rw.updateCh <- &RevisionDestsUpdate{Rev: rw.rev, ClusterIPDest: dest}
```
The caller is going to be the bookkeeping for request count tracking - it needs to know because podIPs need to limit to ContainerConcurrency and clusterIPs need to limit to ContainerConcurrency * replicaCnt.

greghaynes commented on pkg/activator/activator.go:
```diff
@@ -32,6 +32,8 @@ const (
 
 // RevisionID is the combination of namespace and revision name
+// TODO(greghaynes) once we remove old throttler code for revision_backends
+// we can delete this type (in leu of k8s NamespacedName)
```
I think our own type doesnt' really buy us anything as there's no further validation so any meaning is assumed by usage regardless.

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
+		return
+	}
+
+	// If clusterIP is healthy send this update and were done
```
```suggestion
	// If clusterIP is healthy, send this update and we are done.
```

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
+
+	// If clusterIP is healthy send this update and were done
+	if ok, dest, err := rw.probeClusterIP(svc); err != nil {
+		rw.logger.Errorw("Failed to probe clusterIP", zap.Error(err))
```
Do we care to log the ip/name?

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
+
+	// Context used for our probe requests
 	ctx, cancel := context.WithTimeout(context.Background(), probeTimeout)
 	defer cancel()
```
I think having clean separate functions will make reading the code much easier.
Right now it requires lots of concentration :)

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
+	}
+
+	// If clusterIP is healthy send this update and were done
+	if ok, dest, err := rw.probeClusterIP(svc); err != nil {
```
So this, as we know, might take ~10s to complete, due to the iptables reprogramming. This would basically reduce to 0 all the benefits of this work. I think all should be done in parallel and then collected together.

greghaynes replied:
IMO we should set a *very* aggressive timeout here, like 200ms. Assuming I make that change, is there a lot of value in that added complexity?

greghaynes replied:
Actually, I think the context here with timeout will cancel this acter probeTimeout (300ms)?

greghaynes commented on pkg/activator/net/revision_backends.go:
```diff
@@ -119,12 +107,98 @@ type destHealth struct {
 
-	// Context used for our requests
+	if rw.hasHealthyClusterIP() {
+		if !rw.revisionReady() {
```
We need to have a special case for this sending an update when we transition from healthy to unhealthy, but I think youre getting at we can short circuit the rest of this method when we are known dests == 0 which I agree.

greghaynes commented:
I still need to look at using SetupFakeContext but otherwise I think I addressed all comments, and this should be pretty superficial so should be ready for another review from folks.

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+
+	switch len(svcList) {
+	case 0:
+		return nil, fmt.Errorf("Found no private service for revision %q", rw.rev.String())
```
```suggestion
		return nil, fmt.Errorf("found no private service for revision %q", rw.rev.String())
```

greghaynes replied:
Is wanting lowercase for start of an error message actually a thing? If so can we do this via a linter or something.

vagababov replied:
Yes, it is standard Go practice (https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

markusthoemmes replied:
It's also reported by linters IIRC, at least by `staticcheck`.

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+	case 1:
+		return svcList[0], nil
+	default:
+		return nil, fmt.Errorf("Found multiple private services matching revision %v", rw.rev)
```
```suggestion
		return nil, fmt.Errorf("found multiple private services matching revision %v", rw.rev)
```

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+
+func (rw *revisionWatcher) probeClusterIP(svc *corev1.Service) (bool, string, error) {
+	if svc.Spec.ClusterIP == "" {
+		return false, "", fmt.Errorf("Private service %s/%s clusterIP is nil, this should never happen", svc.ObjectMeta.Namespace, svc.ObjectMeta.Name)
```
```suggestion
		return false, "", fmt.Errorf("private service %s/%s clusterIP is nil, this should never happen", svc.ObjectMeta.Namespace, svc.ObjectMeta.Name)
```

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+
+	svcPort, ok := GetServicePort(rw.protocol, svc)
+	if !ok {
+		return false, "", fmt.Errorf("Unable to find HTTP1 port in service %s/%s", svc.Namespace, svc.Name)
```
```suggestion
		return false, "", fmt.Errorf("unable to find HTTP1 port in service %s/%s", svc.Namespace, svc.Name)
```
And elsewhere...

vagababov replied:
Also I think `HTTP1` is no longer relevant?

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+
+		// We know this revision cannot be healthy so short circuit
+		return
+	} else if rw.clusterIPHealthy {
```
else is redundant here, just `if`.

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+
+	destsCh, err := rbm.getOrCreateDestsCh(revID)
+	if err != nil {
+		rbm.logger.Errorw(fmt.Sprintf("Failed to get revision watcher for revision %q", revID.String()),
```
To avoid formatting you can just : `logger.Errorw("Failed...n " + revID.String(), zap...)`.

greghaynes replied:
That would remove quoting around the revision name, IMO its nice to have that grouping.

vagababov commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,414 @@
+	healthStates, err := rw.probePodIPs()
+	if err != nil {
+		rw.logger.Errorw("Failed probing", zap.Error(err))
+		// We dont want to return here as an error still effects health states
```
```suggestion
		// We dont want to return here as an error still affects health states
```

greghaynes commented:
Addressed the last round of comments, can we check out using SetupFakeContext in a later PR if we want to do it - this shouldn't effect correctness and is for setting up test fakes so I think its relatively low cost and im a bit crunched for time while hacking on the throttling code here.

vagababov commented on pkg/activator/testing/roundtripper.go:
```diff
@@ -73,25 +77,37 @@ func response(fr *FakeResponse) (*http.Response, error) {
-	resp := &responses[0]
-	if len(responses) > 1 {
-		rt.ProbeResponses = responses[1:]
+	if _, ok := rt.ProbeHostResponses[host]; ok {
```
```
v, ok := rt.Probe... {
  resp, resps := popRS(v)
  rt.PHR[h] = ...
```

vagababov commented:
Mostly looks good for this iteration
/lgtm

taragu commented on pkg/activator/net/revision_backends.go:
```diff
@@ -0,0 +1,416 @@
+		rw.logger.Errorw(fmt.Sprintf("Failed to probe clusterIP %s/%s", svc.Namespace, svc.Name), zap.Error(err))
+	} else if ok {
+		rw.clusterIPHealthy = true
+		rw.healthStates = make(map[string]bool)
```
I think I'm a bit confused here. Why are we not adding `dest` to `rw.healthStates` here and mark it healthy?

greghaynes replied:
healthStates is only for tracking the health of podIPs, when a clusterIP is healthy (the case in this branch) we assume all addreses in the endpoints object are healthy.

tcnghia commented on pkg/activator/net/helpers.go:
```diff
@@ -0,0 +1,57 @@
+
+// GetServicePort takes a service and a protocol and returns the port number of
+// the port named for that protocol
+func GetServicePort(protocol networking.ProtocolType, svc *corev1.Service) (int, bool) {
```
Name the return value for self-documentation (mainly the `bool`)?

greghaynes commented:
/test pull-knative-serving-integration-tests
/hold cancel

mattmoor commented:
Looks like a flake in a new test?

greghaynes commented:
Yep - I switched the test to use probeHostResponses which doesnt rely on request ordering out of the prober which I suspect was the issue. Ran the test a bunch locally and wasnt able to hit it *shrug*.

greghaynes commented:
ok that obviously wasnt it - I increased the timeout the tests wait for events because I am suspecting we are too aggressive. For sure this is a potential flake source in the TestRevisionBackendsManager test as we are waiting for 200ms probes in some cases. In the last run it was the other test which failed but 100ms might simply be too aggressive here.

greghaynes commented:
This time the flake wasnt me! :)  Looks like TestActivatoinHandler has a data race flake

greghaynes commented:
/test pull-knative-serving-unit-tests
